### ISSUE TITLE
Linear increase in disk-usage for `pandas` objects.

### ISSUE THREAD

**Issue by trendelkampschroer** [joblib/joblib#1316]
I am observing a linear increase in disk-usage for with increasing number of tasks when working on `pandas.DataFrame` objects with `pandas.MultiIndex` indices. The problem disappears when using standard `pd.Index` objects. 

I observe the following numbers for the below example, time is in entire secs and disk usage in 10s or 100s of MB. The "disk-leak" completely disappears when increasing the value of `max_nbytes`. There is a certain performance penalty (in terms of execution time) from increasing `max_nbytes`. 

| # tasks | `max_nbytes` | time (secs.) | disk (MB) |
| ---:       | ---: |       ---: | ---: |
| 100       | 1M  | 8    | 200 |
| 1000    | 1M   | 20 | 3800 |
| 10000  | 1M   | 190 | 38000 |
| 100      | 4M | 6 | 80 |
| 1000   | 4M  | 30 | 80 |
| 10000 | 4M | 260 | 80 |

It would be very convenient if it were possible to set `max_nbytes` via the parallel context , sometimes a user does not have direct access to the `Parallel` object that manages the executions.

Minimal working example below.

```python

import time
from typing import Tuple
import numpy as np
import pandas as pd
from joblib import Parallel, delayed, parallel_backend


def generate_multi_index(n_dates: int, n_entities: int) -> pd.Index:
    dates = pd.date_range("2001-01-01", freq="B", periods=n_dates)
    entities = [f"id_{i}" for i in range(1, n_entities + 1)]
    index = pd.MultiIndex.from_product([dates, entities], names=["date", "entity"])
    return index


def generate_int_index(n_dates: int, n_entities) -> pd.Index:
    return pd.Index(range(n_dates * n_entities))


def generate_str_index(n_dates: int, n_entities) -> pd.Index:
    return pd.Index([f"id_{i}" for i in range(n_dates * n_entities)])


def generate_data(index: pd.Index, n_features: int) -> Tuple[pd.DataFrame, pd.Series]:
    rng = np.random.default_rng()
    columns = [f"feature_{i}" for i in range(n_features)]
    n_samples = len(index)
    X = pd.DataFrame(data=rng.normal(size=(n_samples, n_features)), index=index, columns=columns)
    y = pd.Series(data=rng.normal(size=n_samples), index=index)
    return X, y

def f(X, y):
    time.sleep(0.05)


if __name__ == "__main__":
    n_dates = 2500
    n_entities = 400
    n_features = 10
    index = generate_multi_index(n_dates, n_entities)
    X, y = generate_data(index, n_features)

    with parallel_backend('loky', n_jobs=-1):
        result = Parallel(max_nbytes='1M')([delayed(f)(X, y) for _ in range(1000)])

```

**Comment by ogrisel** [comment]
> Therefore it would be very convenient if it were possible to set max_nbytes via the parallel backend as users do not necessarily have direct access to the Parallel object that manages the executions.

I agree that would be nice. Please consider contributing a pull-request.

As for the original problem, it seems that the multi-index is causing the auto-memmap feature to detect that each parallel task is using a new object instead of reusing the previously memmaped arrays.

I suspect this happens at this line:

https://github.com/joblib/joblib/blob/468a06a709b3b6ee2da95c191933c7e4d4e6a06c/joblib/_memmapping_reducer.py#L399

**Comment by pbcornelius** [comment]
I also have this issue w/ a regular index and a Pandas dataframe. For every run, it re-creates the memmap, which eventually exhausts the shared memory.
